include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -549,6 +547,232 @@ class operator_impl<
+  /**
+   * @brief Inserts a key-value pair `{k, v}` if it's not present in the map. Otherwise, applies
+   * `Op` binary function to the mapped_type corresponding to the key `k` and the value `v`.
+   *
+   * @tparam Value Input type which is implicitly convertible to 'value_type'
+   * @tparam Op Callable type which is used as apply operation and can be
+   *   called with arguments as Op(cuda::atomic_ref<T, Scope>, T). Op strictly must
+   *   have this signature to atomically apply the operation.
+
+   * @param value The element to insert
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and the element to insert.
+   */
+
+  template <typename Value, typename Op>
+  __device__ void insert_or_apply(Value const& value, Op op)
+  {
+    static_assert(cg_size == 1, "Non-CG operation is incompatible with the current probing scheme");
+
+    static_assert(
+      std::is_invocable_v<Op, cuda::atomic_ref<T, Scope>, T>,
+      "insert_or_apply expects `Op` to be a callable as `Op(cuda::atomic_ref<T, Scope>, T)`");
+
+    ref_type& ref_ = static_cast<ref_type&>(*this);
+
+    auto const val         = ref_.impl_.heterogeneous_value(value);
+    auto const key         = ref_.impl_.extract_key(val);
+    auto& probing_scheme   = ref_.impl_.probing_scheme();
+    auto storage_ref       = ref_.impl_.storage_ref();
+    auto probing_iter      = probing_scheme(key, storage_ref.window_extent());
+    auto const empty_value = ref_.impl_.empty_slot_sentinel().second;
+
+    while (true) {
+      auto const window_slots = storage_ref[*probing_iter];
+
+      for (auto& slot_content : window_slots) {
+        auto const eq_res =
+          ref_.impl_.predicate_.operator()<is_insert::YES>(key, slot_content.first);
+        auto const intra_window_index = thrust::distance(window_slots.begin(), &slot_content);
+        auto slot_ptr = (storage_ref.data() + *probing_iter)->data() + intra_window_index;
+
+        // If the key is already in the container, update the payload and return
+        if (eq_res == detail::equal_result::EQUAL) {
+          op(cuda::atomic_ref<T, Scope>{slot_ptr->second}, val.second);
+          return;
+        }
+        if (eq_res == detail::equal_result::AVAILABLE) {
+          switch (ref_.impl_.attempt_insert_stable(slot_ptr, slot_content, val)) {
+            case insert_result::SUCCESS: return;
+            case insert_result::DUPLICATE: {
+              ref_.impl_.wait_for_payload(slot_ptr->second, empty_value);
+              op(cuda::atomic_ref<T, Scope>{slot_ptr->second}, val.second);
+              return;
+            }
+            default: continue;
+          }
+        }
+      }
+      ++probing_iter;
+    }
+  }
+
+  template <typename Value>
+  __device__ void insert_or_apply(Value const& value, cuco::op::reduce::sum_tag)
+  {
+    auto& ref_ = static_cast<ref_type&>(*this);
+    ref_.insert_or_apply(value, [](cuda::atomic_ref<T, Scope> slot_ref, T const& payload) {
```

Comment:
For `atomic_ref` passing via ref, I would prefer passing by value as it would keep the implementation simpler, because it allows r-values from both the user code and in the implementation. Where as passing by ref creates more restriction, even though slightly lightweight.